WIP: Verification

dmorejon opened the pull request:
Parses the baseline file to make sure the test is all PASS. `verifyTransformation` runs the web tests. If the exit status is non-zero, an error is thrown. If the transformation is successful, the baseline file is deleted.

More bug fixes/ features:
Remove `explicit_done: false` option in `setup()` since `single_test: true` mandates `explicit_done: true`.
Add `done();` to end of the last script.
CLI flags added to disable verification and specify a target build
`transformResult` enum return to specify transformation SUCCESS, FAIL, or SKIP.
Summary output for the tool
Concatenate tree.content hack to fix posthtml parser behavior
Babel template placeholder pattern fic
Report untransformed `js-test.js` functions, throw error to propagate failure up the tool.
`run_web_tests.py` flag changes

dmorejon committed:
Add <title>, strip description, debug, index.js add setup once

dmorejon committed:
Added additional documentation, updated index.js transformation.

dmorejon committed:
Updated testTransform to test title return value

Changed transform.js to use the first description() call only.

dmorejon committed:
Indentation Regex Fix

dmorejon committed:
PR Comments: comments, style changes, renaming

dmorejon committed:
PR Comments: documentation, changed logging.

Using npm debug package for logging now, will need to change to
configure the enable/disable of logging in different modules.

dmorejon committed:
ESLint fixes

dmorejon committed:
Only transform .html files, change debug() to console.log()

dmorejon committed:
Remove testRunner.dumpAsText()

dmorejon committed:
Logging with debug(), filter out tests without  js-test

dmorejon committed:
Merge branch 'master' of github.com:Ecosystem-Infra/web-test-transformer into MVP

dmorejon committed:
Modified newline at the end of scripts, logging changes

dmorejon committed:
OutputDir flag, throw error if js-test.js not src'd

dmorejon committed:
Refactored index.js: moved transformFile to own file

dmorejon committed:
transformFile documentation

dmorejon committed:
Parse baseline file

dmorejon committed:
Merge branch 'master' of github.com:Ecosystem-Infra/web-test-transformer into verification

dmorejon committed:
Verification code

Parses baseline file, runs tests, implicitly checking error status with
try/catch.

dmorejon committed:
Updated comments

dmorejon committed:
Minor bug fixes- unescaped period in string and delete baseline file before testing

LukeZielinski commented on src/verify.js:
```diff
@@ -0,0 +1,58 @@
+  let allPass = true;
+  lines.forEach((line) => {
+    if (line.startsWith('FAIL') || line.startsWith('WARN')) {
+      allPass = false;
```
can probably return immediately here?

dmorejon replied:
Hmmm I was confused how to do that since the return would just return within the closure, not the outer function. Right? I could add the return but it would do nothing at the moment. I could also change this to a normal for loop that way I could return from the outer function.

LukeZielinski replied:
Yeah, you're right, doesn't apply to the closure. Normal for-loop with an early return would just be a micro-optimization, so might not be worth the effort. Feel free to ignore.

LukeZielinski commented on src/verify.js:
```diff
@@ -0,0 +1,58 @@
+}
+
+// transformedFile is a path to an html test
+// Assumes cwd is chromium/src/third_party/blink/web_tests/
```
nit: make the tool resilient to these assumptions being broken (ie: throw a meaningful error, or be flexibly about where we run from)

dmorejon replied:
Updated cwd check, is there a good way to make sure that chromium is built beforehand? The command will fail if it doesn't exist, but beyond that not sure what else to do.

LukeZielinski replied:
No real good way to ensure chromium is built, but a failure that clearly tells the user they need to build chromium is fine.

LukeZielinski commented on src/verify.js:
```diff
@@ -0,0 +1,58 @@
+    // this should work. If it is a full path, this should split and take the
+    // end (relative) path.
+    const relativePath = transformedFile.split('web_tests/').pop();
+    const cmd = '../tools/run_web_tests.py -t Default ' + relativePath;
```
the `Default` part should probably be a parameter to this tool - not all engineers call their build "Default"

dmorejon replied:
Outdated, should be updated now

LukeZielinski commented on test/testdata/input/assign-expected.txt:
```diff
@@ -0,0 +1,20 @@
+This test checks whether various forms of assignment expression are allowed.
+
+On success, you will see a series of "PASS" messages, followed by "TEST COMPLETE".
```
I assume these are copied from real baseline files? It'd be nice if the words at the top of each file briefly described what they are, such as "all Pass", "one Fail", "all Pass with Fail inside the message" etc

dmorejon replied:
I was going to do this, but since we are actually parsing all of the file's contents I was concerned about adding anything.

LukeZielinski replied:
sg

LukeZielinski commented on src/verify.js:
```diff
@@ -0,0 +1,58 @@
+  // Delete baseline file if all pass,
+  // testharness.js does not have a baseline if the test is all pass.
+  fs.unlinkSync(baselineFile);
+  try {
```
I could imagine expanding unit tests to cover this try/catch block if we created a dummy py script that mimic'ed the output of `run_web_tests.py`? Not sure how much value that would give us though

Probably not something we want to cover right now anyway, but if you think there's value, lets file a bug.

dmorejon replied:
I do think there's value, I really wanted to test this. A nice refactor would be to pass a path to run_web_tests.py and mock it. I need to look at the --run-web-tests flag (which can take a full path to the test) so we don't need to be in the web_tests/ directory. I think that flag doesn't do the best things with the -expected.txt files, so it needs a bit more research. I'll file an issue with this idea

LukeZielinski replied:
sgtm, add a TODO here with the issue number

dmorejon committed:
Verification features, many bug fixes:

CLI flags added to disable verification and specify a target build
transformResult enum return to specify transformation success,failure,
or skip.
Summary output for the tool
Concatenate tree.content hack to fix posthtml parser behavior
babel template placeholder pattern fic
report untransformed js-test.js functions
Add done(); to end of the last script
run_web_tests.py flag changes

LukeZielinski commented on src/injectScripts.js:
```diff
@@ -112,6 +112,18 @@ function insertTitlePlugin(description) {
 }
 
+
+function concatContentPlugin() {
```
Lets add some comments about why this is necessary

LukeZielinski commented on src/transformFile.js:
```diff
@@ -93,17 +108,21 @@ function transformFile(filePath, outputDir=null) {
-module.exports = {transformFile};
+module.exports = {transformFile, transformResult};
+
+transformFile('./play.html', './play');
```
Is this supposed to be here?

dmorejon replied:
No! Thanks.

LukeZielinski commented on index.js:
```diff
@@ -34,19 +40,43 @@ async function main() {
+    return;
   }
 
   const fileNames = fs.readdirSync(dir);
```
nit: put a comment here saying that the code below is for processing a dir

Alternatively (although low priority, so maybe a TODO is fine), it might be nicer if this code always operated on a list of files, and a helper function was responsible for either finding which files to process inside a dir (and returning a list of them), or just returning a single-element list containing `file`

dmorejon committed:
PR comments: cleanup, comments

dmorejon committed:
PR comments: check cwd assumptions, lint fixes

dmorejon committed:
PR Comments

Hexcles commented on index.js:
```diff
@@ -34,19 +40,48 @@ async function main() {
   });
+  log('Skipped Transformations:', skippedTransformations);
+  log('Failed Transformations:', failedTransformations);
+  log('Failed Verifications:', failedVerifications);
```
What do these messages look like? Do we not need a placeholder in the format string?

dmorejon replied:
Since they are arrays of strings, the output looked fine to me when I tested manually... Is there a better practice?

Hexcles replied:
Oh then it's totally fine.

LukeZielinski commented on src/verify.js:
```diff
@@ -0,0 +1,60 @@
+// Assumes content_shell is built and up to date.
+// Returns true if the transformedFile passses all tests, false otherwise.
+function verifyTransformation(transformedFile, targetBuild) {
+  if (process.cwd().split('/').pop() != 'web_tests') {
```
wonder if this could just be `if (process.cwd().endswith(third_party/blink/web_tests)`

dmorejon committed:
PR comments

Hexcles commented on src/transformScript.js:
```diff
@@ -32,7 +31,8 @@ function transformShouldBeBool() {
           const newActual =
-            babel.template.expression(path.node.arguments[0].value)();
+            babel.template.expression(path.node.arguments[0].value,
+                {placeholderPattern: false})();
```
We've discussed why this is needed offline, but it'd be good not to repeat this everywhere. Could you make a wrapper and add some comment for why this is needed there?

Tangent: I wonder if we can use https://babeljs.io/docs/en/babel-parser#babelparserparseexpressioncode-options instead.

dmorejon replied:
Importing `@babel/parser` and using parseExpression worked, and I think that's better. Replaced everywhere. Thanks for finding this!

Hexcles reviewed (approved):
LGTM % nits

dmorejon committed:
Use babelParser.parseExpression() for evaluating string args
